From 3ff8119ab8f794806a4cc82e21f760c123d92966 Mon Sep 17 00:00:00 2001
From: Magnus Jensen <magnus@fleetdm.com>
Date: Wed, 28 Jan 2026 09:46:53 -0500
Subject: [PATCH] Windows MDM app level impl (#38842)

<!-- Add the related story/sub-task/bug number, like Resolves #123, or
remove if NA -->
**Related issue:** Resolves
https://github.com/fleetdm/confidential/issues/13775

Feature branch merging into main, so all code should be reviewed

---------

Co-authored-by: Sarah Gillespie <73313222+gillespi314@users.noreply.github.com>
---
 pkg/mdm/mdmtest/windows.go              | 207 +++++++++++++--
 server/datastore/mysql/microsoft_mdm.go |  46 +++-
 server/fleet/datastore.go               |   5 +
 server/fleet/microsoft_mdm.go           |  67 +++--
 server/mdm/microsoft/syncml/syncml.go   |   6 +
 server/mock/datastore_mock.go           |  24 ++
 server/mock/redis/key_value_store.go    |  41 +++
 server/mock/service.go                  |   6 +
 server/service/integration_mdm_test.go  | 123 +++------
 server/service/microsoft_mdm.go         | 329 ++++++++++++++++++------
 server/service/microsoft_mdm_test.go    | 185 ++++++++++++-
 11 files changed, 829 insertions(+), 210 deletions(-)
 create mode 100644 server/mock/redis/key_value_store.go

diff --git a/pkg/mdm/mdmtest/windows.go b/pkg/mdm/mdmtest/windows.go
index 6bcdf8a43f2..96ba3ea9e99 100644
--- a/pkg/mdm/mdmtest/windows.go
+++ b/pkg/mdm/mdmtest/windows.go
@@ -2,6 +2,7 @@ package mdmtest
 
 import (
 	"bytes"
+	"crypto/md5" //nolint:gosec // Windows MDM Auth uses MD5
 	"crypto/rsa"
 	"crypto/tls"
 	"encoding/base64"
@@ -49,6 +50,10 @@ type TestWindowsMDMClient struct {
 	jwtSigningKey *rsa.PrivateKey
 	// jwtSigningKeyID is the ID to report in the header for the signing key
 	jwtSigningKeyID string
+
+	username string
+	password string
+	nonce    string
 }
 
 // This is a test-only enrollment type to force erroneous behavior.
@@ -194,29 +199,74 @@ func (c *TestWindowsMDMClient) doManagementReq(rawXMLReq []byte) (map[string]fle
 		fmt.Println(string(rawXMLReq))
 	}
 
-	// TODO: this request works because we're allowing devices without
-	// certificates to communicate with the server. We will need to include the
-	// certificate we generated during enrollment when we fix that.
-	managementResp, err := c.request(microsoft_mdm.MDE2ManagementPath, rawXMLReq)
-	if err != nil {
-		return nil, err
+	sendRequest := func(req []byte) (*fleet.SyncML, error) {
+		managementResp, err := c.request(microsoft_mdm.MDE2ManagementPath, req)
+		if err != nil {
+			return nil, err
+		}
+
+		rawXMLResp, err := io.ReadAll(managementResp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("reading response body: %w", err)
+		}
+
+		if c.debug {
+			fmt.Println("=============== management response ================")
+			fmt.Println(string(rawXMLResp))
+		}
+
+		var syncML fleet.SyncML
+		if err := xml.Unmarshal(rawXMLResp, &syncML); err != nil {
+			return nil, fmt.Errorf("unmarshalling response body: %w", err)
+		}
+
+		return &syncML, nil
 	}
 
-	rawXMLResp, err := io.ReadAll(managementResp.Body)
+	syncML, err := sendRequest(rawXMLReq)
 	if err != nil {
-		return nil, fmt.Errorf("reading response body: %w", err)
+		return nil, err
 	}
 
-	if c.debug {
-		fmt.Println("=============== management response ================")
-		fmt.Println(string(rawXMLResp))
+	if username, password := c.isRekeyRequest(syncML); username != "" && password != "" {
+		c.username = username
+		c.password = password
+
+		// We rekeyed, so we need to resend the original request
+		syncML, err = sendRequest(rawXMLReq)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	var syncML fleet.SyncML
-	if err := xml.Unmarshal(rawXMLResp, &syncML); err != nil {
-		return nil, fmt.Errorf("unmarshalling response body: %w", err)
+	if shouldAuth, nonce := c.shouldAuth(syncML); shouldAuth {
+		var reqSyncML fleet.SyncML
+		if err := xml.Unmarshal(rawXMLReq, &reqSyncML); err != nil {
+			return nil, fmt.Errorf("unmarshalling request body for auth: %w", err)
+		}
+
+		extractedNonce, _ := base64.StdEncoding.DecodeString(*nonce)
+		c.nonce = string(extractedNonce)
+		reqSyncML.SyncHdr.Cred = c.getCredHDR()
+
+		// resend the request but now with credentials
+		rawXMLReq, err = xml.MarshalIndent(reqSyncML, "", "\t")
+		if err != nil {
+			return nil, fmt.Errorf("serializing XML req with auth: %w", err)
+		}
+
+		if c.debug {
+			fmt.Println("=============== management request with auth ================")
+			fmt.Println(string(rawXMLReq))
+		}
+
+		syncML, err = sendRequest(rawXMLReq)
+		if err != nil {
+			return nil, err
+		}
 	}
-	c.lastManagementResp = &syncML
+
+	c.lastManagementResp = syncML
 
 	cmds := make(map[string]fleet.ProtoCmdOperation)
 	for _, p := range c.lastManagementResp.GetOrderedCmds() {
@@ -228,6 +278,26 @@ func (c *TestWindowsMDMClient) doManagementReq(rawXMLReq []byte) (map[string]fle
 	return cmds, nil
 }
 
+func (c *TestWindowsMDMClient) isRekeyRequest(req *fleet.SyncML) (username string, password string) {
+	for _, cmd := range req.GetOrderedCmds() {
+		if cmd.Verb == fleet.CmdReplace && strings.Contains(cmd.Cmd.GetTargetURI(), "AAuthName") {
+			username = cmd.Cmd.GetTargetData()
+		} else if cmd.Verb == fleet.CmdReplace && strings.Contains(cmd.Cmd.GetTargetURI(), "AAuthSecret") {
+			password = cmd.Cmd.GetTargetData()
+		}
+	}
+	return
+}
+
+func (c *TestWindowsMDMClient) shouldAuth(req *fleet.SyncML) (bool, *string) {
+	for _, cmd := range req.GetOrderedCmds() {
+		if cmd.Verb == fleet.CmdStatus && cmd.Cmd.Chal != nil {
+			return true, cmd.Cmd.Chal.Meta.NextNonce.Content
+		}
+	}
+	return false, nil
+}
+
 func (c *TestWindowsMDMClient) SendResponse() (map[string]fleet.ProtoCmdOperation, error) {
 	// Get SessionID
 	sessionID, err := c.lastManagementResp.GetSessionID()
@@ -256,6 +326,7 @@ func (c *TestWindowsMDMClient) SendResponse() (map[string]fleet.ProtoCmdOperatio
 		Target: &fleet.LocURI{
 			LocURI: ptr.String(c.fleetServerURL + microsoft_mdm.MDE2ManagementPath),
 		},
+		Cred: c.getCredHDR(),
 	}
 
 	// iterate over mocked responses and append them to the SyncML message
@@ -271,6 +342,30 @@ func (c *TestWindowsMDMClient) SendResponse() (map[string]fleet.ProtoCmdOperatio
 	return c.doManagementReq(xmlReq)
 }
 
+func (c *TestWindowsMDMClient) getCredHDR() *fleet.CredHdr {
+	return &fleet.CredHdr{
+		Meta: fleet.Meta{
+			Type: &fleet.MetaAttr{
+				XMLNS:   syncml.SyncMLMetaNamespace,
+				Content: ptr.String(syncml.AuthMD5),
+			},
+			Format: &fleet.MetaAttr{
+				XMLNS:   syncml.SyncMLMetaNamespace,
+				Content: ptr.String(syncml.AuthB64Format),
+			},
+		},
+		Data: c.hashedCredentials(),
+	}
+}
+
+func (c *TestWindowsMDMClient) hashedCredentials() string {
+	credentials := fmt.Sprintf("%s:%s", c.username, c.password)
+	credentialsHash := md5.Sum([]byte(credentials)) //nolint:gosec // Windows MDM Auth uses MD5
+	credentialsWithNonce := fmt.Sprintf("%s:%s", base64.StdEncoding.EncodeToString(credentialsHash[:]), c.nonce)
+	digestHash := md5.Sum([]byte(credentialsWithNonce)) //nolint:gosec // Windows MDM Auth uses MD5
+	return base64.StdEncoding.EncodeToString(digestHash[:])
+}
+
 // AppendResponse sets a response for a specific command UUID.
 func (c *TestWindowsMDMClient) AppendResponse(op fleet.SyncMLCmd) {
 	c.queuedCommandResponses[op.CmdID.Value] = op
@@ -403,6 +498,53 @@ YioVozr1IWYySwWVzMf/SUwKZkKJCAJmSVcixE+4kxPkyPGyauIrN3wWC0zb+mjF
 		return fmt.Errorf("enroll request returned SOAP fault: %s", string(body))
 	}
 
+	var soapResponse fleet.SoapResponse
+	if err := xml.Unmarshal(body, &soapResponse); err != nil {
+		return fmt.Errorf("unmarshalling enroll response body: %w", err)
+	}
+
+	decoded, err := base64.StdEncoding.DecodeString(soapResponse.Body.RequestSecurityTokenResponseCollection.RequestSecurityTokenResponse.RequestedSecurityToken.BinarySecurityToken.Content)
+	if err != nil {
+		return fmt.Errorf("decoding enroll response binary security token: %w", err)
+	}
+
+	// strip xml header
+	decoded = bytes.TrimPrefix(decoded, []byte(xml.Header))
+	var provDoc fleet.WapProvisioningDoc
+	if err := xml.Unmarshal(decoded, &provDoc); err != nil {
+		return fmt.Errorf("unmarshalling enroll response provisioning doc: %w", err)
+	}
+
+Outer:
+	for _, char := range provDoc.Characteristics {
+		if char.Type != "APPLICATION" {
+			continue
+		}
+
+		for _, appChar := range char.Characteristics {
+			if appChar.Type != "APPAUTH" {
+				continue
+			}
+			username := ""
+			password := ""
+			for _, appAuthParam := range appChar.Params {
+				if appAuthParam.Name == "AAUTHNAME" {
+					username = appAuthParam.Value
+				}
+				if appAuthParam.Name == "AAUTHSECRET" {
+					password = appAuthParam.Value
+				}
+			}
+
+			// We can do this since only the client credentials characteristic has both username and password, the other one only has password.
+			if username != "" && password != "" {
+				c.username = username
+				c.password = password
+				break Outer
+			}
+		}
+	}
+
 	return nil
 }
 
@@ -578,3 +720,38 @@ func (c *TestWindowsMDMClient) getToken() (binarySecToken string, tokenValueType
 
 	return binarySecToken, tokenValueType, nil
 }
+
+func (c *TestWindowsMDMClient) Unenroll() error {
+	unenrollRequest := []byte(`
+			 <SyncML xmlns="SYNCML:SYNCML1.2">
+			<SyncHdr>
+				<VerDTD>1.2</VerDTD>
+				<VerProto>DM/1.2</VerProto>
+				<SessionID>2</SessionID>
+				<MsgID>1</MsgID>
+				<Target>
+				<LocURI>` + c.fleetServerURL + microsoft_mdm.MDE2ManagementPath + `</LocURI>
+				</Target>
+				<Source>
+				<LocURI>` + c.DeviceID + `</LocURI>
+				</Source>
+			</SyncHdr>
+			<SyncBody>
+				<Alert>
+				<CmdID>4</CmdID>
+				<Data>1226</Data>
+				<Item>
+					<Meta>
+					<Type xmlns="syncml:metinf">com.microsoft:mdm.unenrollment.userrequest</Type>
+					<Format xmlns="syncml:metinf">int</Format>
+					</Meta>
+					<Data>1</Data>
+				</Item>
+				</Alert>
+				<Final/>
+			</SyncBody>
+			</SyncML>`)
+
+	_, err := c.doManagementReq(unenrollRequest)
+	return err
+}
diff --git a/server/datastore/mysql/microsoft_mdm.go b/server/datastore/mysql/microsoft_mdm.go
index ef017f87434..afdccf4c275 100644
--- a/server/datastore/mysql/microsoft_mdm.go
+++ b/server/datastore/mysql/microsoft_mdm.go
@@ -58,6 +58,8 @@ func (ds *Datastore) MDMWindowsGetEnrolledDeviceWithDeviceID(ctx context.Context
 		enroll_proto_version,
 		enroll_client_version,
 		not_in_oobe,
+		credentials_hash,
+		credentials_acknowledged,
 		created_at,
 		updated_at,
 		host_uuid
@@ -89,6 +91,8 @@ func (ds *Datastore) MDMWindowsGetEnrolledDeviceWithHostUUID(ctx context.Context
 		enroll_proto_version,
 		enroll_client_version,
 		not_in_oobe,
+		credentials_hash,
+		credentials_acknowledged,
 		created_at,
 		updated_at,
 		host_uuid
@@ -120,9 +124,11 @@ func (ds *Datastore) MDMWindowsInsertEnrolledDevice(ctx context.Context, device
 			enroll_proto_version,
 			enroll_client_version,
 			not_in_oobe,
-			host_uuid)
+			host_uuid,
+			credentials_hash,
+			credentials_acknowledged)
 		VALUES
-			(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
+			(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
 		ON DUPLICATE KEY UPDATE
 			mdm_device_id         = VALUES(mdm_device_id),
 			device_state          = VALUES(device_state),
@@ -133,7 +139,9 @@ func (ds *Datastore) MDMWindowsInsertEnrolledDevice(ctx context.Context, device
 			enroll_proto_version  = VALUES(enroll_proto_version),
 			enroll_client_version = VALUES(enroll_client_version),
 			not_in_oobe           = VALUES(not_in_oobe),
-			host_uuid             = VALUES(host_uuid)
+			host_uuid             = VALUES(host_uuid),
+			credentials_hash      = VALUES(credentials_hash),
+			credentials_acknowledged = VALUES(credentials_acknowledged)
 	`
 	_, err := ds.writer(ctx).ExecContext(
 		ctx,
@@ -148,7 +156,9 @@ func (ds *Datastore) MDMWindowsInsertEnrolledDevice(ctx context.Context, device
 		device.MDMEnrollProtoVersion,
 		device.MDMEnrollClientVersion,
 		device.MDMNotInOOBE,
-		device.HostUUID)
+		device.HostUUID,
+		device.CredentialsHash,
+		device.CredentialsAcknowledged)
 	if err != nil {
 		if IsDuplicate(err) {
 			return ctxerr.Wrap(ctx, alreadyExists("MDMWindowsEnrolledDevice", device.MDMHardwareID))
@@ -2521,3 +2531,31 @@ func (ds *Datastore) ResendWindowsMDMCommand(ctx context.Context, mdmDeviceId st
 		return nil
 	})
 }
+
+func (ds *Datastore) MDMWindowsUpdateEnrolledDeviceCredentials(ctx context.Context, deviceId string, credentialsHash []byte) error {
+	if deviceId == "" {
+		return nil
+	}
+
+	_, err := ds.writer(ctx).ExecContext(ctx, `
+		UPDATE mdm_windows_enrollments
+		SET credentials_hash = ?
+		WHERE mdm_device_id = ?`,
+		credentialsHash, deviceId,
+	)
+	return err
+}
+
+func (ds *Datastore) MDMWindowsAcknowledgeEnrolledDeviceCredentials(ctx context.Context, deviceId string) error {
+	if deviceId == "" {
+		return nil
+	}
+
+	_, err := ds.writer(ctx).ExecContext(ctx, `
+		UPDATE mdm_windows_enrollments
+		SET credentials_acknowledged = TRUE
+		WHERE mdm_device_id = ?`,
+		deviceId,
+	)
+	return err
+}
diff --git a/server/fleet/datastore.go b/server/fleet/datastore.go
index 564fefe90ad..3bf3d8cc222 100644
--- a/server/fleet/datastore.go
+++ b/server/fleet/datastore.go
@@ -2662,6 +2662,11 @@ type Datastore interface {
 	// RetryVPPInstall retries a single VPP install that failed for the host.
 	// It makes sure to queue a new nano command and update the command_uuid in the host_vpp_software_installs table, as well as the execution ID for the activity.
 	RetryVPPInstall(ctx context.Context, vppInstall *HostVPPSoftwareInstallLite) error
+
+	// MDMWindowsUpdateEnrolledDeviceCredentials updates the credentials hash for the enrolled Windows device.
+	MDMWindowsUpdateEnrolledDeviceCredentials(ctx context.Context, deviceId string, credentialsHash []byte) error
+	// MDMWindowsAcknowledgeEnrolledDeviceCredentials marks the enrolled Windows device credentials as acknowledged.
+	MDMWindowsAcknowledgeEnrolledDeviceCredentials(ctx context.Context, deviceId string) error
 }
 
 type AndroidDatastore interface {
diff --git a/server/fleet/microsoft_mdm.go b/server/fleet/microsoft_mdm.go
index c8054df12a6..141253b8456 100644
--- a/server/fleet/microsoft_mdm.go
+++ b/server/fleet/microsoft_mdm.go
@@ -17,6 +17,7 @@ import (
 
 const (
 	WINDOWS_SCEP_LOC_URI_PART = "/Vendor/MSFT/ClientCertificateInstall/SCEP"
+	WindowsMDMAuthNoncePrefix = "mwenonce:"
 )
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////
@@ -30,12 +31,12 @@ const (
 // SoapResponse is the Soap Envelope Response type for MS-MDE2 responses from the server
 // This envelope XML message is composed by a mandatory SOAP envelope, a SOAP header, and a SOAP body
 type SoapResponse struct {
-	XMLName xml.Name       `xml:"s:Envelope"`
+	XMLName xml.Name       `xml:"http://schemas.xmlsoap.org/soap/envelope/ Envelope"`
 	XMLNSS  string         `xml:"xmlns:s,attr"`
 	XMLNSA  string         `xml:"xmlns:a,attr"`
 	XMLNSU  *string        `xml:"xmlns:u,attr,omitempty"`
-	Header  ResponseHeader `xml:"s:Header"`
-	Body    BodyResponse   `xml:"s:Body"`
+	Header  ResponseHeader `xml:"http://schemas.xmlsoap.org/soap/envelope/ Header"`
+	Body    BodyResponse   `xml:"http://schemas.xmlsoap.org/soap/envelope/ Body"`
 }
 
 // SoapRequest is the Soap Envelope Request type for MS-MDE2 responses to the server
@@ -819,20 +820,22 @@ func (msg WapProvisioningDoc) GetEncodedB64Representation() (string, error) {
 /// Contains the information of the enrolled Windows host
 
 type MDMWindowsEnrolledDevice struct {
-	ID                     uint      `db:"id"`
-	HostUUID               string    `db:"host_uuid"`
-	MDMDeviceID            string    `db:"mdm_device_id"`
-	MDMHardwareID          string    `db:"mdm_hardware_id"`
-	MDMDeviceState         string    `db:"device_state"`
-	MDMDeviceType          string    `db:"device_type"`
-	MDMDeviceName          string    `db:"device_name"`
-	MDMEnrollType          string    `db:"enroll_type"`
-	MDMEnrollUserID        string    `db:"enroll_user_id"`
-	MDMEnrollProtoVersion  string    `db:"enroll_proto_version"`
-	MDMEnrollClientVersion string    `db:"enroll_client_version"`
-	MDMNotInOOBE           bool      `db:"not_in_oobe"`
-	CreatedAt              time.Time `db:"created_at"`
-	UpdatedAt              time.Time `db:"updated_at"`
+	ID                      uint      `db:"id"`
+	HostUUID                string    `db:"host_uuid"`
+	MDMDeviceID             string    `db:"mdm_device_id"`
+	MDMHardwareID           string    `db:"mdm_hardware_id"`
+	MDMDeviceState          string    `db:"device_state"`
+	MDMDeviceType           string    `db:"device_type"`
+	MDMDeviceName           string    `db:"device_name"`
+	MDMEnrollType           string    `db:"enroll_type"`
+	MDMEnrollUserID         string    `db:"enroll_user_id"`
+	MDMEnrollProtoVersion   string    `db:"enroll_proto_version"`
+	MDMEnrollClientVersion  string    `db:"enroll_client_version"`
+	MDMNotInOOBE            bool      `db:"not_in_oobe"`
+	CredentialsHash         *[]byte   `db:"credentials_hash"`
+	CredentialsAcknowledged bool      `db:"credentials_acknowledged"`
+	CreatedAt               time.Time `db:"created_at"`
+	UpdatedAt               time.Time `db:"updated_at"`
 }
 
 func (e MDMWindowsEnrolledDevice) AuthzType() string {
@@ -906,12 +909,18 @@ type SyncHdr struct {
 	Target    *LocURI  `xml:"Target,omitempty"`
 	Source    *LocURI  `xml:"Source,omitempty"`
 	Meta      *MetaHdr `xml:"Meta,omitempty"`
+	Cred      *CredHdr `xml:"Cred,omitempty"`
 }
 
 type MetaHdr struct {
 	MaxMsgSize *string `xml:"MaxMsgSize,omitempty"`
 }
 
+type CredHdr struct {
+	Meta Meta   `xml:"Meta"`
+	Data string `xml:"Data"`
+}
+
 // ProtoCmds contains a slice of SyncML protocol commands
 type ProtoCmds []SyncMLCmd
 
@@ -998,13 +1007,14 @@ func (c *CmdID) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
 
 // Protocol Command
 type SyncMLCmd struct {
-	XMLName xml.Name  `xml:",omitempty"`
-	CmdID   CmdID     `xml:"CmdID"`
-	MsgRef  *string   `xml:"MsgRef,omitempty"`
-	CmdRef  *string   `xml:"CmdRef,omitempty"`
-	Cmd     *string   `xml:"Cmd,omitempty"`
-	Data    *string   `xml:"Data,omitempty"`
-	Items   []CmdItem `xml:"Item,omitempty"`
+	XMLName xml.Name         `xml:",omitempty"`
+	CmdID   CmdID            `xml:"CmdID"`
+	MsgRef  *string          `xml:"MsgRef,omitempty"`
+	CmdRef  *string          `xml:"CmdRef,omitempty"`
+	Cmd     *string          `xml:"Cmd,omitempty"`
+	Data    *string          `xml:"Data,omitempty"`
+	Items   []CmdItem        `xml:"Item,omitempty"`
+	Chal    *SyncMLChallenge `xml:"Chal,omitempty"`
 
 	// ReplaceCommands is a catch-all for any nested <Replace> commands,
 	// which can be found under <Atomic> elements.
@@ -1022,6 +1032,15 @@ type SyncMLCmd struct {
 	ExecCommands []SyncMLCmd `xml:"Exec,omitempty"`
 }
 
+type SyncMLChallenge struct {
+	Meta ChallengeMeta `xml:"Meta"`
+}
+
+type ChallengeMeta struct {
+	Meta
+	NextNonce MetaAttr `xml:"NextNonce,omitempty"`
+}
+
 // ParseWindowsMDMCommand parses the raw XML as a single Windows MDM command.
 // A single <Exec> command is accepted as input.
 func ParseWindowsMDMCommand(rawXMLCmd []byte) (*SyncMLCmd, error) {
diff --git a/server/mdm/microsoft/syncml/syncml.go b/server/mdm/microsoft/syncml/syncml.go
index 57c7abf52bf..c73b2163fae 100644
--- a/server/mdm/microsoft/syncml/syncml.go
+++ b/server/mdm/microsoft/syncml/syncml.go
@@ -108,6 +108,9 @@ const (
 	// This response code will be generated if you try to access a property that the CSP doesn't support
 	CmdStatusOptionalFeature = "406"
 
+	// This response code will be generated if authentication is required to continue the session
+	CmdStatusAuthenticationRequired = "407"
+
 	// Unsupported type or format
 	// This response code can result from XML parsing or formatting errors
 	CmdStatusUnsupportedType = "415"
@@ -294,6 +297,9 @@ const (
 
 	// FleetdWindowsInstallerGUID is the GUID used for fleetd on Windows
 	FleetdWindowsInstallerGUID = "./Device/Vendor/MSFT/EnterpriseDesktopAppManagement/MSI/%7BA427C0AA-E2D5-40DF-ACE8-0D726A6BE096%7D/DownloadInstall"
+
+	AuthMD5       = "syncml:auth-md5"
+	AuthB64Format = "b64"
 )
 
 // MS-MDM Message constants
diff --git a/server/mock/datastore_mock.go b/server/mock/datastore_mock.go
index d9ad5250675..0a005ce7ece 100644
--- a/server/mock/datastore_mock.go
+++ b/server/mock/datastore_mock.go
@@ -1761,6 +1761,10 @@ type GetHostVPPInstallByCommandUUIDFunc func(ctx context.Context, commandUUID st
 
 type RetryVPPInstallFunc func(ctx context.Context, vppInstall *fleet.HostVPPSoftwareInstallLite) error
 
+type MDMWindowsUpdateEnrolledDeviceCredentialsFunc func(ctx context.Context, deviceId string, credentialsHash []byte) error
+
+type MDMWindowsAcknowledgeEnrolledDeviceCredentialsFunc func(ctx context.Context, deviceId string) error
+
 type DataStore struct {
 	AppConfigFunc        AppConfigFunc
 	AppConfigFuncInvoked bool
@@ -4369,6 +4373,12 @@ type DataStore struct {
 	RetryVPPInstallFunc        RetryVPPInstallFunc
 	RetryVPPInstallFuncInvoked bool
 
+	MDMWindowsUpdateEnrolledDeviceCredentialsFunc        MDMWindowsUpdateEnrolledDeviceCredentialsFunc
+	MDMWindowsUpdateEnrolledDeviceCredentialsFuncInvoked bool
+
+	MDMWindowsAcknowledgeEnrolledDeviceCredentialsFunc        MDMWindowsAcknowledgeEnrolledDeviceCredentialsFunc
+	MDMWindowsAcknowledgeEnrolledDeviceCredentialsFuncInvoked bool
+
 	mu sync.Mutex
 }
 
@@ -10454,3 +10464,17 @@ func (s *DataStore) RetryVPPInstall(ctx context.Context, vppInstall *fleet.HostV
 	s.mu.Unlock()
 	return s.RetryVPPInstallFunc(ctx, vppInstall)
 }
+
+func (s *DataStore) MDMWindowsUpdateEnrolledDeviceCredentials(ctx context.Context, deviceId string, credentialsHash []byte) error {
+	s.mu.Lock()
+	s.MDMWindowsUpdateEnrolledDeviceCredentialsFuncInvoked = true
+	s.mu.Unlock()
+	return s.MDMWindowsUpdateEnrolledDeviceCredentialsFunc(ctx, deviceId, credentialsHash)
+}
+
+func (s *DataStore) MDMWindowsAcknowledgeEnrolledDeviceCredentials(ctx context.Context, deviceId string) error {
+	s.mu.Lock()
+	s.MDMWindowsAcknowledgeEnrolledDeviceCredentialsFuncInvoked = true
+	s.mu.Unlock()
+	return s.MDMWindowsAcknowledgeEnrolledDeviceCredentialsFunc(ctx, deviceId)
+}
diff --git a/server/mock/redis/key_value_store.go b/server/mock/redis/key_value_store.go
new file mode 100644
index 00000000000..c83934716bc
--- /dev/null
+++ b/server/mock/redis/key_value_store.go
@@ -0,0 +1,41 @@
+// Automatically generated by mockimpl. DO NOT EDIT!
+
+package mock
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+)
+
+var _ fleet.KeyValueStore = (*KeyValueStore)(nil)
+
+type SetFunc func(ctx context.Context, key string, value string, expireTime time.Duration) error
+
+type GetFunc func(ctx context.Context, key string) (*string, error)
+
+type KeyValueStore struct {
+	SetFunc        SetFunc
+	SetFuncInvoked bool
+
+	GetFunc        GetFunc
+	GetFuncInvoked bool
+
+	mu sync.Mutex
+}
+
+func (kv *KeyValueStore) Set(ctx context.Context, key string, value string, expireTime time.Duration) error {
+	kv.mu.Lock()
+	kv.SetFuncInvoked = true
+	kv.mu.Unlock()
+	return kv.SetFunc(ctx, key, value, expireTime)
+}
+
+func (kv *KeyValueStore) Get(ctx context.Context, key string) (*string, error) {
+	kv.mu.Lock()
+	kv.GetFuncInvoked = true
+	kv.mu.Unlock()
+	return kv.GetFunc(ctx, key)
+}
diff --git a/server/mock/service.go b/server/mock/service.go
index 5c57f202f1c..6ac07e5cda4 100644
--- a/server/mock/service.go
+++ b/server/mock/service.go
@@ -2,9 +2,15 @@ package mock
 
 import (
 	"github.com/fleetdm/fleet/v4/server/fleet"
+	kvmock "github.com/fleetdm/fleet/v4/server/mock/redis"
 	svcmock "github.com/fleetdm/fleet/v4/server/mock/service"
 )
 
 //go:generate go run ./mockimpl/impl.go -o service/service_mock.go "s *Service" "fleet.Service"
+//go:generate go run ./mockimpl/impl.go -o redis/key_value_store.go "kv *KeyValueStore" "fleet.KeyValueStore"
 
 var _ fleet.Service = new(svcmock.Service)
+
+type KVStore struct {
+	kvmock.KeyValueStore
+}
diff --git a/server/service/integration_mdm_test.go b/server/service/integration_mdm_test.go
index f03549b612f..0ebd9236142 100644
--- a/server/service/integration_mdm_test.go
+++ b/server/service/integration_mdm_test.go
@@ -8797,52 +8797,19 @@ func (s *integrationMDMTestSuite) TestWindowsAzureInitiatedEnrollmentAndMapping(
 
 func (s *integrationMDMTestSuite) TestValidManagementUnenrollRequest() {
 	t := s.T()
-
-	// Target Endpoint URL for the management endpoint
-	targetEndpointURL := microsoft_mdm.MDE2ManagementPath
+	ctx := t.Context()
 
 	// Target DeviceID to use
-	deviceID := "DB257C3A08778F4FB61E2749066C1F27"
-
-	// Inserting new device
-	enrolledDevice := &fleet.MDMWindowsEnrolledDevice{
-		MDMDeviceID:            deviceID,
-		MDMHardwareID:          uuid.New().String() + uuid.New().String(),
-		MDMDeviceState:         uuid.New().String(),
-		MDMDeviceType:          "CIMClient_Windows",
-		MDMDeviceName:          "DESKTOP-1C3ARC1",
-		MDMEnrollType:          "ProgrammaticEnrollment",
-		MDMEnrollUserID:        "upn@domain.com",
-		MDMEnrollProtoVersion:  "5.0",
-		MDMEnrollClientVersion: "10.0.19045.2965",
-		MDMNotInOOBE:           false,
-	}
-
-	err := s.ds.MDMWindowsInsertEnrolledDevice(context.Background(), enrolledDevice)
-	require.NoError(t, err)
+	_, mdmHost := createWindowsHostThenEnrollMDM(s.ds, s.server.URL, t)
+	deviceID := mdmHost.DeviceID
 
 	// Checking if device was enrolled
-	_, err = s.ds.MDMWindowsGetEnrolledDeviceWithDeviceID(context.Background(), deviceID)
+	_, err := s.ds.MDMWindowsGetEnrolledDeviceWithDeviceID(ctx, deviceID)
 	require.NoError(t, err)
 
-	// Preparing the SyncML unenroll request
-	requestBytes, err := s.newSyncMLUnenrollMsg(deviceID, targetEndpointURL)
-	require.NoError(t, err)
-
-	resp := s.DoRaw("POST", targetEndpointURL, requestBytes, http.StatusOK)
-
-	// Checking that Command error code was updated
-
-	// Checking response headers
-	require.Contains(t, resp.Header["Content-Type"], syncml.SyncMLContentType)
-
-	// Read response data
-	resBytes, err := io.ReadAll(resp.Body)
+	_, err = mdmHost.StartManagementSession()
 	require.NoError(t, err)
-
-	// Checking if response can be unmarshalled to an golang type
-	var xmlType interface{}
-	err = xml.Unmarshal(resBytes, &xmlType)
+	err = mdmHost.Unenroll()
 	require.NoError(t, err)
 
 	// Checking if device was unenrolled
@@ -10270,56 +10237,6 @@ func (s *integrationMDMTestSuite) newSecurityTokenMsg(encodedBinToken string, de
 	return requestBytes, nil
 }
 
-func (s *integrationMDMTestSuite) newSyncMLUnenrollMsg(deviceID string, managementUrl string) ([]byte, error) {
-	if len(managementUrl) == 0 {
-		return nil, errors.New("managementUrl is empty")
-	}
-
-	return []byte(`
-			 <SyncML xmlns="SYNCML:SYNCML1.2">
-			<SyncHdr>
-				<VerDTD>1.2</VerDTD>
-				<VerProto>DM/1.2</VerProto>
-				<SessionID>2</SessionID>
-				<MsgID>1</MsgID>
-				<Target>
-				<LocURI>` + managementUrl + `</LocURI>
-				</Target>
-				<Source>
-				<LocURI>` + deviceID + `</LocURI>
-				</Source>
-			</SyncHdr>
-			<SyncBody>
-				<Alert>
-				<CmdID>2</CmdID>
-				<Data>1201</Data>
-				</Alert>
-				<Alert>
-				<CmdID>3</CmdID>
-				<Data>1224</Data>
-				<Item>
-					<Meta>
-					<Type xmlns="syncml:metinf">com.microsoft/MDM/LoginStatus</Type>
-					</Meta>
-					<Data>user</Data>
-				</Item>
-				</Alert>
-				<Alert>
-				<CmdID>4</CmdID>
-				<Data>1226</Data>
-				<Item>
-					<Meta>
-					<Type xmlns="syncml:metinf">com.microsoft:mdm.unenrollment.userrequest</Type>
-					<Format xmlns="syncml:metinf">int</Format>
-					</Meta>
-					<Data>1</Data>
-				</Item>
-				</Alert>
-				<Final/>
-			</SyncBody>
-			</SyncML>`), nil
-}
-
 func (s *integrationMDMTestSuite) checkMDMProfilesSummaries(t *testing.T, teamID *uint, expectedSummary fleet.MDMProfilesSummary, expectedAppleSummary *fleet.MDMProfilesSummary) {
 	var queryParams []string
 	if teamID != nil {
@@ -20429,3 +20346,31 @@ func (s *integrationMDMTestSuite) TestInstalledApplicationListCommandForBYODiDev
 	checkExpectedCommands(mdmClientBYOD, true, 1)
 	checkExpectedCommands(mdmClientDEP, false, 1)
 }
+
+func (s *integrationMDMTestSuite) TestWindowsRekeyFlow() {
+	t := s.T()
+	ctx := t.Context()
+
+	_, mdmHost := createWindowsHostThenEnrollMDM(s.ds, s.server.URL, t)
+
+	// Now we remove the credentials_hash and ack to simulate an existing enrollment to force rekeying
+	mysql.ExecAdhocSQL(t, s.ds, func(q sqlx.ExtContext) error {
+		_, err := q.ExecContext(ctx, "UPDATE mdm_windows_enrollments SET credentials_hash = NULL, credentials_acknowledged = FALSE WHERE mdm_device_id = ?", mdmHost.DeviceID)
+		return err
+	})
+
+	_, err := mdmHost.StartManagementSession()
+	require.NoError(t, err)
+
+	var updatedValues struct {
+		CredentialsHash         *[]byte `db:"credentials_hash"`
+		CredentialsAcknowledged bool    `db:"credentials_acknowledged"`
+	}
+	mysql.ExecAdhocSQL(t, s.ds, func(q sqlx.ExtContext) error {
+		err := sqlx.GetContext(ctx, q, &updatedValues, "SELECT credentials_hash, credentials_acknowledged FROM mdm_windows_enrollments WHERE mdm_device_id = ?", mdmHost.DeviceID)
+		return err
+	})
+
+	require.NotNil(t, updatedValues.CredentialsHash)
+	require.True(t, updatedValues.CredentialsAcknowledged)
+}
diff --git a/server/service/microsoft_mdm.go b/server/service/microsoft_mdm.go
index 4ac341a27ef..aa1ad2a4f8b 100644
--- a/server/service/microsoft_mdm.go
+++ b/server/service/microsoft_mdm.go
@@ -3,6 +3,7 @@ package service
 import (
 	"bytes"
 	"context"
+	"crypto/md5" //nolint:gosec // Windows MDM Auth uses MD5
 	"crypto/x509"
 	"database/sql"
 	"encoding/base64"
@@ -645,7 +646,7 @@ func isEligibleForWindowsMDMMigration(host *fleet.Host, mdmInfo *fleet.HostMDM)
 // The Application Provisioning configuration is used for bootstrapping a device with an OMA DM account
 // The paramenters here maps to the W7 application CSP
 // https://learn.microsoft.com/en-us/windows/client-management/mdm/w7-application-csp
-func NewApplicationProvisioningData(mdmEndpoint string) mdm_types.Characteristic {
+func NewApplicationProvisioningData(mdmEndpoint string, username string, secret string) mdm_types.Characteristic {
 	provDoc := newCharacteristic("APPLICATION", []mdm_types.Param{
 		// The PROVIDER-ID parameter specifies the server identifier for a management server used in the current management session
 		newParm("PROVIDER-ID", syncml.DocProvisioningAppProviderID, ""),
@@ -689,9 +690,9 @@ func NewApplicationProvisioningData(mdmEndpoint string) mdm_types.Characteristic
 			newParm("AAUTHLEVEL", "APPSRV", ""),
 			// DIGEST - Specifies that the SyncML DM 'syncml:auth-md5' authentication type.
 			newParm("AAUTHTYPE", "DIGEST", ""),
-			newParm("AAUTHNAME", "dummy", ""),
-			newParm("AAUTHSECRET", "dummy", ""),
-			newParm("AAUTHDATA", "nonce", ""),
+			newParm("AAUTHNAME", username, ""),
+			newParm("AAUTHSECRET", secret, ""),
+			newParm("AAUTHDATA", "nonce", ""), // We don't care about setting the first round nonce, as when the device checks in we will prompt the credentials and pass a new nonce.
 		}, nil),
 	})
 
@@ -1190,7 +1191,7 @@ func (svc *Service) GetMDMWindowsEnrollResponse(ctx context.Context, secTokenMsg
 	}
 
 	// Getting the device provisioning information in the form of a WapProvisioningDoc
-	deviceProvisioning, err := svc.getDeviceProvisioningInformation(ctx, secTokenMsg)
+	deviceProvisioning, credentialsHash, err := svc.getDeviceProvisioningInformation(ctx, secTokenMsg)
 	if err != nil {
 		return nil, ctxerr.Wrap(ctx, err, "device provisioning information")
 	}
@@ -1213,7 +1214,7 @@ func (svc *Service) GetMDMWindowsEnrollResponse(ctx context.Context, secTokenMsg
 	//
 	// This method also creates the relevant enrollment activity as it has
 	// access to the device information.
-	err = svc.storeWindowsMDMEnrolledDevice(ctx, userID, hostUUID, secTokenMsg)
+	err = svc.storeWindowsMDMEnrolledDevice(ctx, userID, hostUUID, secTokenMsg, credentialsHash)
 	if err != nil {
 		return nil, ctxerr.Wrap(ctx, err, "enrolled device information cannot be stored")
 	}
@@ -1228,20 +1229,25 @@ func (svc *Service) GetMDMWindowsManagementResponse(ctx context.Context, reqSync
 	}
 
 	// Checking if the incoming request is trusted
-	err := svc.isTrustedRequest(ctx, reqSyncML, reqCerts)
+	requestAuthState, err := svc.isTrustedRequest(ctx, reqSyncML, reqCerts)
 	if err != nil {
 		return nil, ctxerr.Wrap(ctx, err, "management request is not trusted")
 	}
 
+	// Token is authorized
+	svc.authz.SkipAuthorization(ctx)
+
+	if requestAuthState == RequestAuthStateRekey {
+		// If we signalled to rekey the device, we short-circuit into a rekey flow.
+		return svc.rekeyWindowsDevice(ctx, reqSyncML)
+	}
+
 	// Getting the management response message
-	resSyncMLmsg, err := svc.getManagementResponse(ctx, reqSyncML)
+	resSyncMLmsg, err := svc.getManagementResponse(ctx, reqSyncML, requestAuthState)
 	if err != nil {
 		return nil, ctxerr.Wrap(ctx, err, "management response message")
 	}
 
-	// Token is authorized
-	svc.authz.SkipAuthorization(ctx)
-
 	return resSyncMLmsg, nil
 }
 
@@ -1264,59 +1270,156 @@ func (svc *Service) GetMDMWindowsTOSContent(ctx context.Context, redirectUri str
 	return htmlBuf.String(), nil
 }
 
+type requestAuthState int
+
+const (
+	RequestAuthStateUntrusted requestAuthState = iota
+	RequestAuthStateUnauthorized
+	RequestAuthStateChallenge
+	RequestAuthStateRekey
+	RequestAuthStateTrusted
+)
+
 // isTrustedRequest checks if the incoming request was sent from MDM enrolled device
-func (svc *Service) isTrustedRequest(ctx context.Context, reqSyncML *fleet.SyncML, reqCerts []*x509.Certificate) error {
+// It returns a boolean if we should challenge the device and an error if the request/device calling is not trusted
+func (svc *Service) isTrustedRequest(ctx context.Context, reqSyncML *fleet.SyncML, reqCerts []*x509.Certificate) (requestAuthState, error) {
 	if reqSyncML == nil {
-		return fleet.NewInvalidArgumentError("syncml req message", "message is not present")
+		return RequestAuthStateUntrusted, fleet.NewInvalidArgumentError("syncml req message", "message is not present")
 	}
 
 	// Checking if calling request is coming from an already MDM enrolled device
 	deviceID, err := reqSyncML.GetSource()
 	if err != nil || deviceID == "" {
-		return fmt.Errorf("invalid SyncML message %w", err)
+		return RequestAuthStateUntrusted, fmt.Errorf("invalid SyncML message %w", err)
 	}
 
 	enrolledDevice, err := svc.ds.MDMWindowsGetEnrolledDeviceWithDeviceID(ctx, deviceID)
 	if err != nil || enrolledDevice == nil {
-		return errors.New("device was not MDM enrolled")
+		return RequestAuthStateUntrusted, errors.New("device was not MDM enrolled")
 	}
 
 	// Check if TLS certs contains device ID on its common name
 	if len(reqCerts) > 0 {
 		for _, reqCert := range reqCerts {
 			if strings.Contains(reqCert.Subject.CommonName, deviceID) {
-				return nil
+				return RequestAuthStateTrusted, nil
 			}
 		}
 	}
 
-	// TODO: Latest version of the MDM client stack don't populate TLS.PeerCertificates array
-	// This is a temporary workaround to allow the management request to proceed
-	// Transport-level security should be replaced for Application-level security
-	// Transport-level security is defined in the MS-MDM spec in section 1.3.1
-	// On the other hand, Application-level security is defined here
-	// https://www.openmobilealliance.org/release/DM/V1_2_1-20080617-A/OMA-TS-DM_Security-V1_2_1-20080617-A.pdf
-	// The initial values for Application-level security configuration are defined in the
-	// WAP Profile blob that is sent to the device during the enrollment process. Example below
-	//	<characteristic type="APPAUTH">
-	//		<parm name="AAUTHLEVEL" value="CLIENT"/>
-	//		<parm name="AAUTHTYPE" value="DIGEST"/>
-	//		<parm name="AAUTHSECRET" value="2jsidqgffx"/>
-	//		<parm name="AAUTHDATA" value="aGVsbG8gd29ybGQ="/>
-	//	</characteristic>
-	//	<characteristic type="APPAUTH">
-	//		<parm name="AAUTHLEVEL" value="APPSRV"/>
-	//		<parm name="AAUTHTYPE" value="DIGEST"/>
-	//		<parm name="AAUTHNAME" value="43f8bf591b8557346021"/>
-	//		<parm name="AAUTHSECRET" value="crbr3w2cab"/>
-	//		<parm name="AAUTHDATA" value="aGVsbG8gd29ybGQ="/>
-	//	</characteristic>
-
-	if len(reqCerts) == 0 {
-		return nil
+	if !enrolledDevice.CredentialsAcknowledged && enrolledDevice.CredentialsHash == nil {
+		// Device has not gotten new credentials, rekey the device only once
+		return RequestAuthStateRekey, nil
+	}
+
+	if reqSyncML.SyncHdr.Cred == nil {
+		// No certs, but no credentials present - challenge the device
+		return RequestAuthStateChallenge, nil
+	}
+
+	// Extract the last nonce used to generate the credentials hash
+	nonce, err := svc.keyValueStore.Get(ctx, fleet.WindowsMDMAuthNoncePrefix+deviceID)
+	if err != nil {
+		return RequestAuthStateUntrusted, ctxerr.Wrap(ctx, err, "get device nonce from kv store")
+	}
+
+	if nonce == nil || *nonce == "" {
+		// Challenge the device if nonce is missing, which will send a new nonce and store it
+		return RequestAuthStateChallenge, nil
+	}
+
+	// Credentials are present, validate it
+	credFormat := reqSyncML.SyncHdr.Cred.Meta.Format
+	credType := reqSyncML.SyncHdr.Cred.Meta.Type
+	credData := reqSyncML.SyncHdr.Cred.Data
+
+	if credFormat == nil || credType == nil || credFormat.Content == nil || credType.Content == nil {
+		return RequestAuthStateUntrusted, errors.New("SyncML credentials format or type is missing")
+	}
+
+	if *credFormat.Content != syncml.AuthB64Format || *credType.Content != syncml.AuthMD5 {
+		return RequestAuthStateUntrusted, errors.New("SyncML credentials format or type is invalid")
 	}
 
-	return errors.New("calling device is not trusted")
+	// MD5 auth digest, which includes (username:password):nonce
+	// Where username:password is hashed and b64 encoded, and then further hased with the nonce and finally b64 encoded for transport
+	// https://www.openmobilealliance.org/release/DM/V1_2_1-20080617-A/OMA-TS-DM_Security-V1_2_1-20080617-A.pdf Chaper (5.3)
+	receivedDigestHash, err := base64.StdEncoding.DecodeString(credData)
+	if err != nil {
+		return RequestAuthStateUntrusted, ctxerr.Wrap(ctx, err, "decode SyncML credentials data")
+	}
+
+	encodedCredentialsHash := base64.StdEncoding.EncodeToString(*enrolledDevice.CredentialsHash)
+	expectedDigest := fmt.Sprintf("%s:%s", encodedCredentialsHash, *nonce)
+	expectedDigestHash := md5.Sum([]byte(expectedDigest)) //nolint:gosec // Windows MDM Auth uses MD5
+
+	if !bytes.Equal(receivedDigestHash, expectedDigestHash[:]) {
+		// Credentials do not match what we expect
+		return RequestAuthStateUnauthorized, nil
+	}
+
+	// We verified the username, password and nonce match what we expect, so we can ack the rekeyed credentials
+	if !enrolledDevice.CredentialsAcknowledged {
+		err = svc.ds.MDMWindowsAcknowledgeEnrolledDeviceCredentials(ctx, enrolledDevice.MDMDeviceID)
+		if err != nil {
+			return RequestAuthStateUntrusted, ctxerr.Wrap(ctx, err, "mark device credentials as acknowledged")
+		}
+	}
+
+	return RequestAuthStateTrusted, nil
+}
+
+func (svc *Service) rekeyWindowsDevice(ctx context.Context, reqSyncML *fleet.SyncML) (*fleet.SyncML, error) {
+	if reqSyncML == nil {
+		return nil, fleet.NewInvalidArgumentError("syncml req message", "message is not present")
+	}
+
+	// Getting the device ID from the SyncML message
+	deviceID, err := reqSyncML.GetSource()
+	if err != nil || deviceID == "" {
+		return nil, fmt.Errorf("invalid SyncML message %w", err)
+	}
+
+	enrolledDevice, err := svc.ds.MDMWindowsGetEnrolledDeviceWithDeviceID(ctx, deviceID)
+	if err != nil || enrolledDevice == nil {
+		return nil, errors.New("device was not MDM enrolled")
+	}
+
+	username := deviceID
+	password := uuid.NewString()
+	credentialsHash := md5.Sum(fmt.Appendf([]byte{}, "%s:%s", username, password)) //nolint:gosec // Windows MDM Auth uses MD5
+
+	// Store the new credentials hash and mark that the device has not acknowledged them yet
+	err = svc.ds.MDMWindowsUpdateEnrolledDeviceCredentials(ctx, enrolledDevice.MDMDeviceID, credentialsHash[:])
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "update enrolled device credentials")
+	}
+
+	// Queue two replace commands to update the credentials on the device
+	accountUid := "0x0000800cABF8CA7D87C0FA21BB21B896AE2C8468CA07710326415228ACF2392EA1327077"
+	usernameReplace := newSyncMLCmdText("Replace", fmt.Sprintf("./SyncML/DMAcc/%s/AppAuth/CLCRED/AAuthName", accountUid), username)
+	usernameReplace.CmdID = mdm_types.CmdID{
+		IncludeFleetComment: true,
+		Value:               "rekey-credentials-username",
+	}
+	passwordReplace := newSyncMLCmdText("Replace", fmt.Sprintf("./SyncML/DMAcc/%s/AppAuth/CLCRED/AAuthSecret", accountUid), password)
+	passwordReplace.CmdID = mdm_types.CmdID{
+		IncludeFleetComment: true,
+		Value:               "rekey-credentials-password",
+	}
+
+	// Get the incoming MessageID
+	reqMessageID, err := reqSyncML.GetMessageID()
+	if err != nil {
+		return nil, fmt.Errorf("get incoming msg: %w", err)
+	}
+
+	// We only create a response here, and does not persist it to the DB to avoid saving the secrets in plain text
+	return svc.createResponseSyncML(ctx, reqSyncML, []*mdm_types.SyncMLCmd{
+		NewSyncMLCmdStatus(reqMessageID, "0", syncml.SyncMLHdrName, syncml.CmdStatusOK), // We need to ack the incoming message to send rekey commands
+		usernameReplace,
+		passwordReplace,
+	})
 }
 
 // isFleetdPresentOnDevice checks if the device requires Fleetd to be deployed
@@ -1523,18 +1626,75 @@ func (svc *Service) processIncomingAlertsCommands(ctx context.Context, messageID
 
 // processIncomingMDMCmds process the incoming message from the device
 // It will return the list of operations that need to be sent to the device
-func (svc *Service) processIncomingMDMCmds(ctx context.Context, deviceID string, reqMsg *fleet.SyncML) ([]*fleet.SyncMLCmd, error) {
+func (svc *Service) processIncomingMDMCmds(ctx context.Context, deviceID string, reqMsg *fleet.SyncML, requestAuthState requestAuthState) ([]*fleet.SyncMLCmd, error) {
 	var responseCmds []*fleet.SyncMLCmd
 
+	saveResponse := func(topLevelExists []string) error {
+		enrichedSyncML := fleet.NewEnrichedSyncML(reqMsg)
+		if enrichedSyncML.HasCommands() {
+			if err := svc.ds.MDMWindowsSaveResponse(ctx, deviceID, enrichedSyncML, topLevelExists); err != nil {
+				return fmt.Errorf("store incoming msgs: %w", err)
+			}
+		}
+		return nil
+	}
+
 	// Get the incoming MessageID
 	reqMessageID, err := reqMsg.GetMessageID()
 	if err != nil {
 		return nil, fmt.Errorf("get incoming msg: %w", err)
 	}
 
+	if requestAuthState == RequestAuthStateChallenge || requestAuthState == RequestAuthStateUnauthorized {
+		nonce := uuid.NewString() // using UUID as nonce since it has 122 bits of entropy
+		base64Nonce := base64.StdEncoding.EncodeToString([]byte(nonce))
+		err := svc.keyValueStore.Set(ctx, fleet.WindowsMDMAuthNoncePrefix+deviceID, nonce, 5*time.Minute)
+		if err != nil {
+			return nil, ctxerr.Wrap(ctx, err, "store device nonce in kv store")
+		}
+
+		status := syncml.CmdStatusAuthenticationRequired
+		if requestAuthState == RequestAuthStateUnauthorized {
+			status = syncml.CmdStatusInvalidCredentials
+		}
+
+		ackMsg := NewSyncMLCmdStatus(reqMessageID, "0", syncml.SyncMLHdrName, status)
+		ackMsg.Chal = &fleet.SyncMLChallenge{
+			Meta: fleet.ChallengeMeta{
+				NextNonce: fleet.MetaAttr{
+					XMLNS:   syncml.SyncMLMetaNamespace,
+					Content: &base64Nonce,
+				},
+				Meta: fleet.Meta{
+					Type: &fleet.MetaAttr{
+						XMLNS:   syncml.SyncMLMetaNamespace,
+						Content: ptr.String(syncml.AuthMD5),
+					},
+					Format: &fleet.MetaAttr{
+						XMLNS:   syncml.SyncMLMetaNamespace,
+						Content: ptr.String(syncml.AuthB64Format),
+					},
+				},
+			},
+		}
+
+		responseCmds = append(responseCmds, ackMsg)
+		err = saveResponse([]string{})
+		if err != nil {
+			return nil, err
+		}
+		return responseCmds, nil
+	}
+
+	if requestAuthState != RequestAuthStateTrusted {
+		return nil, errors.New("untrusted request cannot be processed")
+	}
+
 	// Acknowledge the message header
 	// msgref is always 0 for the header
 	if err = reqMsg.IsValidHeader(); err == nil {
+		// We always return 200 here, we could also return 212 to indicate we don't need the credentials every time,
+		// but our logic is built around it being present on each request
 		ackMsg := NewSyncMLCmdStatus(reqMessageID, "0", syncml.SyncMLHdrName, syncml.CmdStatusOK)
 		responseCmds = append(responseCmds, ackMsg)
 	}
@@ -1575,11 +1735,9 @@ func (svc *Service) processIncomingMDMCmds(ctx context.Context, deviceID string,
 		return nil, err
 	}
 
-	enrichedSyncML := fleet.NewEnrichedSyncML(reqMsg)
-	if enrichedSyncML.HasCommands() {
-		if err := svc.ds.MDMWindowsSaveResponse(ctx, deviceID, enrichedSyncML, topLevelExists); err != nil {
-			return nil, fmt.Errorf("store incoming msgs: %w", err)
-		}
+	err = saveResponse(topLevelExists)
+	if err != nil {
+		return nil, err
 	}
 
 	return responseCmds, nil
@@ -1696,7 +1854,7 @@ func (svc *Service) createResponseSyncML(ctx context.Context, req *fleet.SyncML,
 }
 
 // getManagementResponse returns a valid SyncML response message
-func (svc *Service) getManagementResponse(ctx context.Context, reqMsg *fleet.SyncML) (*mdm_types.SyncML, error) {
+func (svc *Service) getManagementResponse(ctx context.Context, reqMsg *fleet.SyncML, requestAuthState requestAuthState) (*mdm_types.SyncML, error) {
 	if reqMsg == nil {
 		return nil, fleet.NewInvalidArgumentError("syncml req message", "message is not present")
 	}
@@ -1708,15 +1866,20 @@ func (svc *Service) getManagementResponse(ctx context.Context, reqMsg *fleet.Syn
 	}
 
 	// Process the incoming MDM protocol commands and get the response MDM protocol commands
-	resIncomingCmds, err := svc.processIncomingMDMCmds(ctx, deviceID, reqMsg)
+	resIncomingCmds, err := svc.processIncomingMDMCmds(ctx, deviceID, reqMsg, requestAuthState)
 	if err != nil {
 		return nil, fmt.Errorf("message processing error %w", err)
 	}
 
-	// Process the pending operations and get the MDM response protocol commands
-	resPendingCmds, err := svc.getPendingMDMCmds(ctx, deviceID)
-	if err != nil {
-		return nil, fmt.Errorf("message processing error %w", err)
+	resPendingCmds := []*mdm_types.SyncMLCmd{}
+
+	if requestAuthState == RequestAuthStateTrusted {
+		// Process the pending operations and get the MDM response protocol commands
+		pendingCmds, err := svc.getPendingMDMCmds(ctx, deviceID)
+		if err != nil {
+			return nil, fmt.Errorf("message processing error %w", err)
+		}
+		resPendingCmds = pendingCmds
 	}
 
 	// Create the response SyncML message
@@ -1754,41 +1917,46 @@ func (svc *Service) removeWindowsDeviceIfAlreadyMDMEnrolled(ctx context.Context,
 // This information is used to configure the device management client
 // See section 2.2.9.1 for more details on the XML provision schema used here
 // https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-mde2/35e1aca6-1b8a-48ba-bbc0-23af5d46907a
-func (svc *Service) getDeviceProvisioningInformation(ctx context.Context, secTokenMsg *fleet.RequestSecurityToken) (string, error) {
+func (svc *Service) getDeviceProvisioningInformation(ctx context.Context, secTokenMsg *fleet.RequestSecurityToken) (string, []byte, error) {
+	reqDeviceID, err := GetContextItem(secTokenMsg, syncml.ReqSecTokenContextItemDeviceID)
+	if err != nil {
+		return "", nil, err
+	}
+
 	// Getting the HW DeviceID from the RequestSecurityToken msg
 	reqHWDeviceID, err := GetContextItem(secTokenMsg, syncml.ReqSecTokenContextItemHWDevID)
 	if err != nil {
-		return "", err
+		return "", nil, err
 	}
 
 	// Getting the EnrollmentType information from the RequestSecurityToken msg
 	reqEnrollType, err := GetContextItem(secTokenMsg, syncml.ReqSecTokenContextItemEnrollmentType)
 	if err != nil {
-		return "", err
+		return "", nil, err
 	}
 
 	// Getting the BinarySecurityToken from the RequestSecurityToken msg
 	binSecurityTokenData, err := secTokenMsg.GetBinarySecurityTokenData()
 	if err != nil {
-		return "", err
+		return "", nil, err
 	}
 
 	// Getting the BinarySecurityToken type from the RequestSecurityToken msg
 	binSecurityTokenType, err := secTokenMsg.GetBinarySecurityTokenType()
 	if err != nil {
-		return "", err
+		return "", nil, err
 	}
 
 	// Getting the client CSR request from the device
 	clientCSR, err := microsoft_mdm.GetClientCSR(binSecurityTokenData, binSecurityTokenType)
 	if err != nil {
-		return "", err
+		return "", nil, err
 	}
 
 	// Getting the signed, DER-encoded certificate bytes and its uppercased, hex-endcoded SHA1 fingerprint
 	rawSignedCertDER, rawSignedCertFingerprint, err := svc.SignMDMMicrosoftClientCSR(ctx, reqHWDeviceID, clientCSR)
 	if err != nil {
-		return "", err
+		return "", nil, err
 	}
 
 	// Preparing client certificate and identity certificate information to be sent to the Windows MDM Enrollment Client
@@ -1802,17 +1970,22 @@ func (svc *Service) getDeviceProvisioningInformation(ctx context.Context, secTok
 	// Preparing the provisioning information that includes the location of the Device Management Service (DMS)
 	appCfg, err := svc.ds.AppConfig(ctx)
 	if err != nil {
-		return "", err
+		return "", nil, err
 	}
 
 	// Getting the MS-MDM management URL to provision the device
 	urlManagementEndpoint, err := microsoft_mdm.ResolveWindowsMDMManagement(appCfg.ServerSettings.ServerURL)
 	if err != nil {
-		return "", err
+		return "", nil, err
 	}
 
+	// generate username and password for device management service
+	username := reqDeviceID
+	password := uuid.NewString()
+	credentialsHash := md5.Sum(fmt.Appendf(nil, "%s:%s", username, password)) //nolint:gosec // Windows MDM Auth uses MD5
+
 	// Preparing the Application Provisioning information
-	appConfigProvisioningData := NewApplicationProvisioningData(urlManagementEndpoint)
+	appConfigProvisioningData := NewApplicationProvisioningData(urlManagementEndpoint, username, password)
 
 	// Preparing the DM Client Provisioning information
 	appDMClientProvisioningData := NewDMClientProvisioningData()
@@ -1821,14 +1994,14 @@ func (svc *Service) getDeviceProvisioningInformation(ctx context.Context, secTok
 	provDoc := NewProvisioningDoc(certStoreProvisioningData, appConfigProvisioningData, appDMClientProvisioningData)
 	encodedProvDoc, err := provDoc.GetEncodedB64Representation()
 	if err != nil {
-		return "", err
+		return "", nil, err
 	}
 
-	return encodedProvDoc, nil
+	return encodedProvDoc, credentialsHash[:], nil
 }
 
 // storeWindowsMDMEnrolledDevice stores the device information to the list of MDM enrolled devices
-func (svc *Service) storeWindowsMDMEnrolledDevice(ctx context.Context, userID string, hostUUID string, secTokenMsg *fleet.RequestSecurityToken) error {
+func (svc *Service) storeWindowsMDMEnrolledDevice(ctx context.Context, userID string, hostUUID string, secTokenMsg *fleet.RequestSecurityToken, credentialsHash []byte) error {
 	const (
 		error_tag = "windows MDM enrolled storage: "
 	)
@@ -1886,17 +2059,19 @@ func (svc *Service) storeWindowsMDMEnrolledDevice(ctx context.Context, userID st
 
 	// Getting the Windows Enrolled Device Information
 	enrolledDevice := &fleet.MDMWindowsEnrolledDevice{
-		MDMDeviceID:            reqDeviceID,
-		MDMHardwareID:          reqHWDevID,
-		MDMDeviceState:         microsoft_mdm.MDMDeviceStateEnrolled,
-		MDMDeviceType:          reqDeviceType,
-		MDMDeviceName:          reqDeviceName,
-		MDMEnrollType:          reqEnrollType,
-		MDMEnrollUserID:        userID, // This could be Host UUID or UPN email
-		MDMEnrollProtoVersion:  reqEnrollVersion,
-		MDMEnrollClientVersion: reqAppVersion,
-		MDMNotInOOBE:           reqNotInOOBE,
-		HostUUID:               hostUUID,
+		MDMDeviceID:             reqDeviceID,
+		MDMHardwareID:           reqHWDevID,
+		MDMDeviceState:          microsoft_mdm.MDMDeviceStateEnrolled,
+		MDMDeviceType:           reqDeviceType,
+		MDMDeviceName:           reqDeviceName,
+		MDMEnrollType:           reqEnrollType,
+		MDMEnrollUserID:         userID, // This could be Host UUID or UPN email
+		MDMEnrollProtoVersion:   reqEnrollVersion,
+		MDMEnrollClientVersion:  reqAppVersion,
+		MDMNotInOOBE:            reqNotInOOBE,
+		HostUUID:                hostUUID,
+		CredentialsHash:         &credentialsHash,
+		CredentialsAcknowledged: true,
 	}
 
 	if err := svc.ds.MDMWindowsInsertEnrolledDevice(ctx, enrolledDevice); err != nil {
diff --git a/server/service/microsoft_mdm_test.go b/server/service/microsoft_mdm_test.go
index 85fb3d6b3d0..b88dd8c9bd6 100644
--- a/server/service/microsoft_mdm_test.go
+++ b/server/service/microsoft_mdm_test.go
@@ -2,11 +2,15 @@ package service
 
 import (
 	"context"
+	"crypto/md5" //nolint:gosec // Windows MDM Auth uses MD5
+	"crypto/x509"
+	"encoding/base64"
 	"encoding/xml"
 	"errors"
 	"fmt"
 	"strings"
 	"testing"
+	"time"
 
 	"github.com/fleetdm/fleet/v4/server/contexts/license"
 	"github.com/fleetdm/fleet/v4/server/fleet"
@@ -287,7 +291,7 @@ func TestProvisioningDocGeneration(t *testing.T) {
 		[]byte{0x4, 0x5, 0x6})
 
 	// Preparing the WAP Provisioning Doc response
-	appConfigData := NewApplicationProvisioningData(microsoft_mdm.MDE2EnrollPath)
+	appConfigData := NewApplicationProvisioningData(microsoft_mdm.MDE2EnrollPath, "testuser", "testpassword")
 	appDMClientData := NewDMClientProvisioningData()
 	provDoc := NewProvisioningDoc(certStoreData, appConfigData, appDMClientData)
 
@@ -297,6 +301,8 @@ func TestProvisioningDocGeneration(t *testing.T) {
 	require.Contains(t, string(outXML), deviceIdentityFingerprint)
 	require.Contains(t, string(outXML), serverIdentityFingerprint)
 	require.Contains(t, string(outXML), microsoft_mdm.MDE2EnrollPath)
+	require.Contains(t, string(outXML), "testuser")
+	require.Contains(t, string(outXML), "testpassword")
 }
 
 func TestValidSyncMLCmdStatus(t *testing.T) {
@@ -910,3 +916,180 @@ func TestReconcileWindowsProfilesWithOneHostFailingStillAddsManagedCertificate(t
 	}
 	require.EqualValues(t, 1, foundErrors, "Should have found one failed status update")
 }
+
+func TestRekeyWindowsDevice(t *testing.T) {
+	ds := new(mock.Store)
+	kv := new(mock.KVStore)
+	svc, ctx := newTestService(t, ds, nil, nil, &TestServerOpts{
+		KeyValueStore: kv,
+	})
+
+	var credsHash *[]byte
+	ds.MDMWindowsGetEnrolledDeviceWithDeviceIDFunc = func(ctx context.Context, mdmDeviceID string) (*fleet.MDMWindowsEnrolledDevice, error) {
+		return &fleet.MDMWindowsEnrolledDevice{
+			MDMDeviceID:     "device",
+			HostUUID:        "host-uuid-123",
+			CredentialsHash: credsHash,
+		}, nil
+	}
+
+	ds.MDMWindowsUpdateEnrolledDeviceCredentialsFunc = func(ctx context.Context, deviceId string, credentialsHash []byte) error {
+		require.Equal(t, "device", deviceId)
+		credsHash = &credentialsHash
+		return nil
+	}
+
+	ackCalled := 0
+	ds.MDMWindowsAcknowledgeEnrolledDeviceCredentialsFunc = func(ctx context.Context, deviceId string) error {
+		require.Equal(t, "device", deviceId)
+		ackCalled++
+		return nil
+	}
+
+	kv.SetFunc = func(ctx context.Context, key string, value string, expireTime time.Duration) error {
+		return nil
+	}
+
+	var nonce string
+	kv.GetFunc = func(ctx context.Context, key string) (*string, error) {
+		return &nonce, nil
+	}
+
+	ds.AppConfigFunc = func(ctx context.Context) (*fleet.AppConfig, error) {
+		return &fleet.AppConfig{
+			ServerSettings: fleet.ServerSettings{
+				ServerURL: "fake-mdm-server.com",
+			},
+		}, nil
+	}
+
+	syncml := `<SyncML xmlns="SYNCML:SYNCML1.2">
+  <SyncHdr>
+    <VerDTD>1.2</VerDTD>
+    <VerProto>DM/1.2</VerProto>
+    <SessionID>1</SessionID>
+    <MsgID>1</MsgID>
+    <Target>
+      <LocURI>fake-mdm-server.com</LocURI>
+    </Target>
+    <Source>
+      <LocURI>device</LocURI>
+    </Source>
+  </SyncHdr>
+  <SyncBody>
+    <Alert>
+      <CmdID>2</CmdID>
+      <Data>1201</Data>
+    </Alert>
+    <Final />
+  </SyncBody>
+</SyncML>`
+
+	var req *fleet.SyncML
+	err := xml.Unmarshal([]byte(syncml), &req)
+	require.NoError(t, err)
+
+	res, err := svc.GetMDMWindowsManagementResponse(ctx, req, []*x509.Certificate{})
+	require.NoError(t, err)
+	require.NotNil(t, res)
+
+	seenStatuses := 0
+	seenReplaces := 0
+	seenOther := 0
+	var username string
+	var password string
+	for _, cmd := range res.SyncBody.Raw {
+		switch cmd.XMLName.Local {
+		case fleet.CmdStatus:
+			require.Equal(t, "200", *cmd.Data)
+			seenStatuses++
+		case fleet.CmdReplace:
+			containsAuthReplace := strings.Contains(cmd.GetTargetURI(), "AAuthName") || strings.Contains(cmd.GetTargetURI(), "AAuthSecret")
+			require.True(t, containsAuthReplace, "Replace command should be for AAuthName or AAuthSecret")
+			seenReplaces++
+
+			if strings.Contains(cmd.GetTargetURI(), "AAuthName") {
+				username = cmd.GetTargetData()
+			} else if strings.Contains(cmd.GetTargetURI(), "AAuthSecret") {
+				password = cmd.GetTargetData()
+			}
+		default:
+			seenOther++
+		}
+	}
+
+	assert.Equal(t, 1, seenStatuses, "should have one Status command")
+	assert.Equal(t, 2, seenReplaces, "should have two Replace commands")
+	assert.Equal(t, 0, seenOther, "should not have other commands")
+
+	// Respond with no credentials again to get a nonce
+	res, err = svc.GetMDMWindowsManagementResponse(ctx, req, []*x509.Certificate{})
+	require.NoError(t, err)
+	require.NotNil(t, res)
+
+	// Require Chal in header
+	require.Len(t, res.SyncBody.Raw, 1, "should short circuit with challenge")
+	chalFound := false
+	for _, cmd := range res.SyncBody.Raw {
+		if cmd.Chal != nil {
+			chalFound = true
+			nonce = *cmd.Chal.Meta.NextNonce.Content
+			break
+		}
+	}
+	require.True(t, chalFound, "should have challenge command")
+
+	// Now respond with credentials to ack the rekey
+	// WE only need to mock this as we short-circuit when challenging or invalid creds
+	ds.MDMWindowsGetPendingCommandsFunc = func(ctx context.Context, deviceID string) ([]*fleet.MDMWindowsCommand, error) {
+		return []*fleet.MDMWindowsCommand{}, nil
+	}
+	ds.GetWindowsMDMCommandsForResendingFunc = func(ctx context.Context, failedCommandIds []string) ([]*fleet.MDMWindowsCommand, error) {
+		return []*fleet.MDMWindowsCommand{}, nil
+	}
+
+	deviceCredsHash := hashMDMCredentials(username, password, nonce)
+	syncmlWithCreds := fmt.Sprintf(`<SyncML xmlns="SYNCML:SYNCML1.2">
+  <SyncHdr>
+    <VerDTD>1.2</VerDTD>
+    <VerProto>DM/1.2</VerProto>
+    <SessionID>1</SessionID>
+    <MsgID>1</MsgID>
+    <Target>
+      <LocURI>fake-mdm-server.com</LocURI>
+    </Target>
+    <Source>
+      <LocURI>device</LocURI>
+    </Source>
+	<Cred>
+		<Meta>
+        <Format xmlns="syncml:metinf">b64</Format>
+        <Type xmlns="syncml:metinf">syncml:auth-md5</Type>
+      </Meta>
+      <Data>%s</Data>
+	</Cred>
+  </SyncHdr>
+  <SyncBody>
+    <Alert>
+      <CmdID>2</CmdID>
+      <Data>1201</Data>
+    </Alert>
+    <Final />
+  </SyncBody>
+</SyncML>`, base64.StdEncoding.EncodeToString(deviceCredsHash))
+	err = xml.Unmarshal([]byte(syncmlWithCreds), &req)
+	require.NoError(t, err)
+
+	res, err = svc.GetMDMWindowsManagementResponse(ctx, req, []*x509.Certificate{})
+	require.NoError(t, err)
+	require.NotNil(t, res)
+
+	require.Equal(t, 1, ackCalled, "acknowledge should have been called once")
+}
+
+func hashMDMCredentials(username, password, nonce string) []byte {
+	credsHash := md5.Sum([]byte(username + ":" + password)) //nolint:gosec // Windows MDM Auth uses MD5
+	encodedCreds := base64.StdEncoding.EncodeToString(credsHash[:])
+	nonceHash := md5.Sum([]byte(encodedCreds + ":" + nonce)) //nolint:gosec // Windows MDM Auth uses MD5
+	return nonceHash[:]
+}
